ingest/src/SnootyDataSource.ts

```diff
@@ -0,0 +1,147 @@
+import fs from "fs";
+import { createInterface } from "readline";
+import { DataSource } from "./DataSource";
+import { Page } from "./updatePages";
+import { snootyAstToMd } from "./snootyAstToMd";
+
+// These types are what's in the snooty manifest jsonl file.
+export type SnootyTimestampEntry = {
+  type: "timestamp";
+  data: number;
+};
+
+export type SnootyMetadataEntry = {
+  type: "metadata";
+  data: Record<string, unknown>;
+};
+
+// TODO
+export type SnootyAsset = Record<string, unknown>;
+
+export type SnootyNode = {
+  type: string;
+  children?: (SnootyNode | SnootyTextNode)[];
+  [key: string]: unknown;
+};
+
+export type SnootyTextNode = SnootyNode & {
+  type: "text";
+  children: never;
+  value: string;
+};
+
+export type SnootyPage = {
+  page_id: string;
+  ast: SnootyNode;
+};
+
+export type SnootyPageEntry = {
+  type: "page";
+  data: SnootyPage;
+};
+
+export type SnootyAssetEntry = {
+  type: "asset";
+  data: Record<string, unknown>;
+};
+
+export const makeSnootyDataSource = async ({
+  name,
+  manifestUrl,
+  baseUrl,
+}: {
+  name: string;
+  manifestUrl: string;
+
+  /**
+    The base url used in links etc
+   */
+  baseUrl: string;
+}): Promise<DataSource> => {
+  const fetch =
+    new URL(manifestUrl).protocol === "file:"
+      ? async (url: string) => ({
+          body: fs.createReadStream(new URL(url).pathname),
+        })
+      : (await import("node-fetch")).default;
```

Comment:
funky! per prev convos, accepting this works even though i don't fully understand what's happening here.

Reply:
ESM bullshit. However, it turns out I can actually fix how ESMs are loaded in our tsconfig so we don't have to load this dynamically.